### DISCUSSION TITLE
Suggestion: @turf/boolean-covers

### DISCUSSION THREAD

**Discussion by rdmurphy** [Turfjs/turf#2790] [Feature proposals]
Hello! Not sure what the best way to approach making a suggestion is, so I'll lay it out here.

Today I was trying to figure out if there was a way to do something like [`ST_Covers`](http://postgis.net/docs/manual-1.4/ST_Covers.html) using Turf. After unsuccessfully attempting to get the desired result using each of the boolean* methods, I finally figured out a way to get what I wanted by combining a few pieces of what Turf currently provides.

```js
const {
  bbox,
  bboxPolygon,
  booleanContains,
  booleanOverlap,
} = require('@turf/turf');

const geoJSONA = {
  /* A */
};
const geoJSONB = {
  /* B */
};

// Bcause we only care about detecting when at least one point of B is
// within A, we can do all our calculations using just the bounds, which
// greatly simplifies things/speeds it up
const geoJSONABBox = bboxPolygon(bbox(geoJSONA));
const geoJSONBBBox = bboxPolygon(bbox(geoJSONB));

// as far as I can tell, the logic of `cover` is accomplished by combining
// the checks of booleanOverlap and booleanContains, I'm not sure if there is
// a more performant way of doing this
if (
  booleanOverlap(geoJSONABBox, geoJSONBBBox) ||
  booleanContains(geoJSONABBox, geoJSONBBBox)
) {
  return true;
} else {
  return false;
}
```

When I attempted to use only `booleanOverlap`, I wouldn't get any features entirely within my collection A. When I attempted to use only `booleanContains`, I wouldn't get any features that only touched collection A. With cover, you want *both*.

Of course, I could just keep doing what I have here and happily keep getting work done. 😄 But this does kind of feel like something that could be in Turf proper!

(There's also the inverse of `ST_Cover` — [`ST_CoveredBy`](http://postgis.net/docs/manual-1.4/ST_CoveredBy.html). Much like `booleanWithin` and `booleanContains`, these two could exist as a pair.)

**Comment by rowanwins** [comment]
Hi @rdmurphy 

Thanks for the suggestion. I've slowly been adding more of the boolean style modules so I'll add `covers` to the to-do list. I think a bbox comparison is part of the answer but I don't think it'll work in all situations. Anyway I'll look into it and try and keep this ticket updated.

Cheers

**Comment by dpmcmlxxvi** [comment]
@rowanwins @stebogit I recently published [de9im](https://github.com/dpmcmlxxvi/de9im) a Javascript library that Turf might find helpful. It implements all of the DE-9IM spatial predicates and I believe addresses many of the issues in the various `turf-boolean` methods. The library uses Turf for low-level spatial operations but otherwise implements the predicates from scratch.

I'd be willing to work on collaborating to integrate it in Turf if there is any interest. It's been a while since I contributed to Turf. After looking at the issue tracker it looks like it is probably best to wait until v7 is released.
